File: plugins/thoughtspot-to-sigma/skills/thoughtspot-to-sigma/scripts/lib/code_rep.py
"""Shape adapter for the Sigma WORKBOOK code representation
(POST /v2/workbooks/spec, GET|PUT /v2/workbooks/{id}/spec, POST /v2/workbooks/spec/verify).

Verified live 2026-08-03/04: this surface nests non-metadata fields under a top-level
`document` key and rejects the old flat body with HTTP 400 — including on /verify.
Sigma engineering confirmed 2026-08-03 that the DATA-MODEL code-rep surface is NOT
changing, so this adapter is workbook-only and always writes the nested shape.
Do NOT use it on /v2/dataModels/.../spec payloads — that API ignores `document`.

Reads stay tolerant of the legacy flat shape because flat artifacts still exist
on disk (committed workbook snapshots, fixtures).
"""

# `settings` (theme/navigation) and `agents` belong INSIDE `document` too — omitting
# them sweeps themeName/themeOverrides/agents onto the top level, where they are not
# valid keys, silently dropping theme + agents on every write.
DOC_KEYS = ("schemaVersion", "pages", "kind", "layout", "settings", "agents")


def document(response):
    """Return the workbook document from either the nested or legacy flat shape."""
    if not isinstance(response, dict):
        return {}
    inner = response.get("document")
    if isinstance(inner, dict):
        return inner
    return {k: v for k, v in response.items() if k in DOC_KEYS}


def metadata(response):
    if not isinstance(response, dict):
        return {}
    return {k: v for k, v in response.items() if k != "document" and k not in DOC_KEYS}


def wrap(doc, extra=None):
    """Build a request body. Every live workbook code-rep endpoint requires the wrapper."""
    out = dict(extra or {})
    out["document"] = doc
    return out


def theme_settings(name=None, overrides=None):
    """The ONE place that knows where theming lives.

    `themeName` / `themeOverrides` were REMOVED from the workbook spec. Theming
    is now `document.settings.theme.{name,overrides}`. Live-probed 2026-08-06:
    `document.themeName` is a hard 400 ("no longer supported. Use
    document.settings.theme.name instead"), and a themeName that lands at the
    TOP level (outside `document`) is accepted and SILENTLY IGNORED — the
    workbook is created unthemed with no error. Emitters must route through
    here rather than hand-rolling either key.

    Returns {} when there is nothing to say, so it is safe to merge blindly.
    """
    theme = {}
    if name:
        theme["name"] = name
    if overrides:
        theme["overrides"] = overrides
    if not theme:
        return {}
    return {"settings": {"theme": theme}}


def merge_settings(doc, fragment):
    """Deep-merge a {'settings': ...} fragment without clobbering siblings."""
    if not fragment:
        return doc
    out = dict(doc)
    for section, value in (fragment.get("settings") or {}).items():
        existing = (out.get("settings") or {}).get(section)
        merged = {**existing, **value} if isinstance(existing, dict) and isinstance(value, dict) else value
        out["settings"] = {**(out.get("settings") or {}), section: merged}
    return out

File: plugins/thoughtspot-to-sigma/skills/thoughtspot-to-sigma/scripts/lib/code_rep.mjs
// Shape adapter for the Sigma WORKBOOK code representation.
// Verified live 2026-08-03/04: nested `document` required on write (including
// /v2/workbooks/spec/verify); flat bodies 400. The DATA-MODEL code-rep surface is
// NOT changing — do not use this on /v2/dataModels/.../spec payloads.

// `settings` (theme/navigation) and `agents` belong INSIDE `document` too — omitting
// them sweeps themeName/themeOverrides/agents onto the top level, where they are not
// valid keys, silently dropping theme + agents on every write.
export const DOC_KEYS = ['schemaVersion', 'pages', 'kind', 'layout', 'settings', 'agents'];

// REMOVED from the API. The workbook theme is now settings.theme.name /
// settings.theme.overrides (published OpenAPI: createWorkbookSpec has zero
// occurrences of themeName/themeOverrides). The individual override keys are
// unchanged - only the container path moved. document() folds the legacy pair
// forward so specs and fixtures written before the move still produce a valid body.
export const LEGACY_THEME_KEYS = ['themeName', 'themeOverrides'];

const isObj = (v) => v !== null && typeof v === 'object' && !Array.isArray(v);

export function document(response) {
  if (!isObj(response)) return {};
  const doc = isObj(response.document)
    ? response.document
    : Object.fromEntries(Object.entries(response).filter(([k]) => DOC_KEYS.includes(k)));
  return foldLegacyTheme(doc, response);
}

// themeName/themeOverrides -> settings.theme.{name,overrides}. Returns the input
// untouched when no legacy key is present (the common, already-correct path).
function foldLegacyTheme(doc, source) {
  const name = doc.themeName || source.themeName;
  const overrides = doc.themeOverrides || source.themeOverrides;
  const hasOv = isObj(overrides) && Object.keys(overrides).length > 0;
  const hasLegacyKey = LEGACY_THEME_KEYS.some((k) => k in doc);
  if (!name && !hasOv && !hasLegacyKey) return doc;

  const out = Object.fromEntries(
    Object.entries(doc).filter(([k]) => !LEGACY_THEME_KEYS.includes(k)),
  );
  const settings = { ...(out.settings || {}) };
  const theme = { ...(settings.theme || {}) };
  if (name && !theme.name) theme.name = name;
  if (hasOv) theme.overrides = { ...(theme.overrides || {}), ...overrides };
  if (Object.keys(theme).length === 0) return out;
  settings.theme = theme;
  out.settings = settings;
  return out;
}

// Emitter helper - set the workbook theme in the CURRENT shape. Builders should
// call this instead of assigning the removed themeName/themeOverrides pair.
export function setTheme(doc, { name = null, overrides = null } = {}) {
  const hasOv = isObj(overrides) && Object.keys(overrides).length > 0;
  if (!name && !hasOv) return doc;
  doc.settings = doc.settings || {};
  doc.settings.theme = doc.settings.theme || {};
  if (name) doc.settings.theme.name = name;
  if (hasOv) {
    doc.settings.theme.overrides = { ...(doc.settings.theme.overrides || {}), ...overrides };
  }
  return doc;
}

// Read the theme from either shape.
export function theme(spec) {
  const t = (document(spec).settings || {}).theme || {};
  return { name: t.name ?? null, overrides: t.overrides || {} };
}

export function metadata(response) {
  if (!isObj(response)) return {};
  return Object.fromEntries(
    Object.entries(response).filter(
      ([k]) => k !== 'document' && !DOC_KEYS.includes(k) && !LEGACY_THEME_KEYS.includes(k),
    ),
  );
}

export function wrap(doc, extra = {}) {
  return { ...extra, document: doc };
}
